include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp

```diff
@@ -0,0 +1,372 @@
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
+ * \brief Provides the algorithm to search in an index using optimum search schemes.
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include <range/v3/view/slice.hpp>
+
+#include <seqan3/search/algorithm/detail/search_common.hpp>
+#include <seqan3/search/algorithm/detail/search_scheme_precomputed.hpp>
+#include <seqan3/search/fm_index/concept.hpp>
+
+/*!\addtogroup search
+ * \{
+ */
+
+namespace seqan3::detail
+{
+
+//!\brief Computes a (not optimal) search scheme.
+inline auto compute_search_scheme(uint8_t const min_error, uint8_t const max_error)
+{
+    // TODO: Simple backtracking. Replace this at least by the pigeonhole principle or even better by 01*0 schemes.
+    std::vector<search_dyn> scheme{{{1}, {min_error}, {max_error}}};
+    return scheme;
+}
+
+//!\brief Returns for each search the cumulative length of blocks in the order of blocks in each search.
+inline auto search_scheme_block_info(auto const & search_scheme, uint64_t const query_length)
```

Comment:
We could but
* search schemes only occur in detail namespace
* the compiler will build a lot of code paths with search schemes (0-4 errors, different combinations of configuration objects, etc.)

IMO this would only increase the compilation time and doesn't pay off just to require the parameter `search_scheme` to fulfill the `search_scheme_concept`.